src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentNodeWriter.cs

```diff
@@ -132,131 +131,253 @@ public override void WriteComponentTypeInferenceMethod(CodeRenderingContext cont
                 {
-                    if (child is ComponentAttributeIntermediateNode attribute)
-                    {
-                        string typeName;
-                        if (attribute.GloballyQualifiedTypeName != null)
-                        {
-                            typeName = attribute.GloballyQualifiedTypeName;
-                        }
-                        else
-                        {
-                            typeName = attribute.TypeName;
-                            if (attribute.BoundAttribute != null && !attribute.BoundAttribute.IsGenericTypedProperty())
-                            {
-                                typeName = "global::" + typeName;
-                            }
-                        }
-
-                        p.Add(($"__seq{p.Count}", typeName, $"__arg{p.Count}"));
-                    }
-                    else if (child is SplatIntermediateNode splat)
-                    {
-                        var typeName = ComponentsApi.AddMultipleAttributesTypeFullName;
-                        p.Add(($"__seq{p.Count}", typeName, $"__arg{p.Count}"));
-                    }
+                    p.Insert(i, new TypeInferenceMethodParameter(null, cascadingGenericType.ValueType, $"__syntheticArg{i}", usedForTypeInference: true, cascadingGenericType));
+                    i++;
                 }
+            }
+
+            return p;
+        }
 
-                foreach (var childContent in node.Component.ChildContents)
+        protected static void TrackCapturedCascadingGenericParameterVariable(ComponentIntermediateNode node, TypeInferenceMethodParameter parameter, string variableName)
+        {
+            // If this captured variable corresponds to a generic type we want to cascade to
+            // descendants, supply that info to descendants
+            if (node.ProvidesCascadingGenericTypes != null)
+            {
+                foreach (var cascadeGeneric in node.ProvidesCascadingGenericTypes.Values)
                 {
-                    var typeName = childContent.TypeName;
-                    if (childContent.BoundAttribute != null && !childContent.BoundAttribute.IsGenericTypedProperty())
+                    if (cascadeGeneric.ValueSourceNode == parameter.Source)
                     {
-                        typeName = "global::" + typeName;
+                        cascadeGeneric.ValueExpression = variableName;
                     }
-                    p.Add(($"__seq{p.Count}", typeName, $"__arg{p.Count}"));
                 }
+            }
 
-                foreach (var capture in node.Component.SetKeys)
-                {
-                    p.Add(($"__seq{p.Count}", "object", $"__arg{p.Count}"));
-                }
+            // Since we've now evaluated and captured this expression, use the variable
+            // instead of the expression from now on
+            parameter.ReplaceSourceWithCapturedVariable(variableName);
```

Comment:
Am I correct in understanding that `ReplaceSourceWithCapturedVariable` is responsible for doing the work of setting the `typeInferenceValue0` variable wherever `GetItems` would have been referenced?

Reply:
Exactly. We're replacing our record of what expression will be emitted with the name of the new variable that holds that expression result.